app/src/main/java/app/revanced/manager/domain/installer/RootInstaller.kt

```diff
@@ -0,0 +1,134 @@
+            Shell.cmd("mount | grep \"$it\"").exec().isSuccess
+        } ?: false
+    }
+
+    fun mount(packageName: String) {
+        if (isAppMounted(packageName)) return
+
+        val stockAPK = pm.getPackageInfo(packageName)?.applicationInfo?.sourceDir
+            ?: throw Exception("Failed to load application info")
+        val patchedAPK = "$modulesPath/$packageName-revanced/$packageName.apk"
+
+        Shell.cmd("mount -o bind \"$patchedAPK\" \"$stockAPK\"").exec()
+            .also { if (!it.isSuccess) throw Exception("Failed to mount APK") }
+    }
+
+    fun unmount(packageName: String) {
+        if (!isAppMounted(packageName)) return
+
+        val stockAPK = pm.getPackageInfo(packageName)?.applicationInfo?.sourceDir
+            ?: throw Exception("Failed to load application info")
+
+        Shell.cmd("umount -l \"$stockAPK\"").exec()
+            .also { if (!it.isSuccess) throw Exception("Failed to unmount APK") }
+    }
+
+    suspend fun install(
+        patchedAPK: File,
+        stockAPK: File,
+        packageName: String,
+        version: String,
+        label: String
+    ) {
+        withContext(Dispatchers.IO) {
+            rootConnection.remoteFS?.let { remoteFS ->
+                val assets = app.assets
+                val modulePath = "$modulesPath/$packageName-revanced"
+
+                unmount(packageName)
+
+                pm.getPackageInfo(packageName)?.let { packageInfo ->
+                    if (packageInfo.versionName <= version)
+                        Shell.cmd("pm uninstall -k --user 0 $packageName").exec()
+                            .also { if (!it.isSuccess) throw Exception("Failed to uninstall stock app") }
+                }
+
+                Shell.cmd("pm install \"${stockAPK.absolutePath}\"").exec()
+                    .also { if (!it.isSuccess) throw Exception("Failed to install stock app") }
+
+                remoteFS.getFile(modulePath).mkdir()
+
+                listOf(
+                    "service.sh",
+                    "module.prop",
+                ).forEach { file ->
+                    assets.open("root/$file").use { inputStream ->
+                        remoteFS.getFile("$modulePath/$file").newOutputStream()
+                            .use { outputStream ->
+                                val buffer = ByteArray(1024)
+                                var bytesRead: Int
+
+                                while (inputStream.read(buffer).also { bytesRead = it } != -1) {
+                                    val content = String(buffer, 0, bytesRead)
+                                        .replace("__PKG_NAME__", packageName)
+                                        .replace("__VERSION__", version)
+                                        .replace("__LABEL__", label)
+                                        .toByteArray()
+
+                                    outputStream.write(content)
+                                }
```

Comment:
This will cause problems in the future if the file is changed in such a way that `__PKG_NAME__` ends up at a 1024-byte boundary, because it will get cut off and not get replaced by the expected value.

Can be solved by reading the entire file instead of streaming (or by using something thats made for streams). Memory needs to be traded for correctness in this situation